### ISSUE TITLE
dtype of ComposedLinearOperator is None

### ISSUE THREAD

**Issue by e-moral-sanchez** [pyccel/psydac#336]
- The `dtype` property is hardcoded in the class `ComposedLinearOperator` to return `None`. 

- `.dtype()` should return complex if any of the operators in the composition has a complex `dtype`

**Comment by jowezarek** [comment]
I agree! I hardcoded `None` because back then complex was not implemented and @yguclu and I were unsure of the purpose of `dtype`, if I remember correctly.
We should then probably apply the same change to `ZeroOperator, IdentityOperator, ScaledLinearOperator, SumLinearOperator, PowerLinearOperator, InverseLinearOperator` and the to-be-renamed class `LinearSolver`.

`ScaledLinearOperator` and `PowerLinearOperator` should inherit the `dtype` from the respective linear operator.

Also, in `LinearOperator.__mul__` we should differentiate between real and complex scalars, the current implementation might cause some bugs.

**Comment by yguclu** [comment]
Maybe `dtype` could specify the type of the entries of the matrix form of the linear operator. That is, the result of `toarray()` should be a 2D NumPy array with `dtype=dtype`. Does that make sense to you?

**Comment by e-moral-sanchez** [comment]
Yes, this is one possibility. I do not have any application in mind, so I am not sure this is the best definition.
